### ISSUE TITLE
Asynchronous API guidelines / standards

### ISSUE THREAD

**Issue by benvolioT** [rackerlabs/riss#105]
Eric Dunham has some ideas for a new set of guidelines / standards for asynchronous API interactions:

These are my thoughts on possible standards/guidelines:

*. Sync/blocking is the default behavior unless otherwise specified in the
documentation (like running a tftp config backup on a device, for example)
*. A user SHOULD be able to specify `respond-async` or `wait` in the
`Prefer` header of the request (per http://tools.ietf.org/html/rfc7240)
*. All requests MUST respond to the HTTP verbs HEAD, OPTIONS, and GET
*. A successful sync GET/POST/PUT/DELETE MUST adhere to
http://tools.ietf.org/html/rfc7231 (yes, I know this bullet is lazy)
*. A successful async GET/POST/PUT/DELETE MUST return a 202 Accepted,
along with a Location header specifying an intermediate resource to check
the status of the operation
*. A successful async GET/POST/PUT/DELETE MAY return a body to indicate
information about the operation, but that body MUST NOT be used in lieu of
the below status codes and headers
*. An intermediate status resource MAY support the HTTP verb DELETE, if
the operation is cancelable
*. An intermediate status resource HEAD/GET MUST be synchronous
*. An intermediate status resource HEAD/GET MUST return a 200 OK if the
operation has not yet completed
*. An intermediate status resource HEAD/GET MAY return a Retry-After
header if the operation has not yet completed
*. An intermediate status resource HEAD/GET MUST return a 303 See Other if
the operation has completed, along with a Location header specifying the
canonical URI of the resource
*. An intermediate status resource HEAD/GET MAY return a 410 Gone after an
arbitrarily long period of time if the operation has been completed

**Comment by benvolioT** [comment]
From @mmorga:

There's a long thread here where Mark Nottingham and Roy Fielding chimed
in on 202.

http://lists.w3.org/Archives/Public/ietf-http-wg/2011OctDec/0076.html


TL;DR.

Req: POST ...
Resp: 202 Accepted
  Headers:
    Link: <http://example.org/status-monitor-url>;rel="monitor"
    Retry-After: 123

GET http://example.org/status-monitor-url
Would return the same response until things were done, when you'd get:

Resp: 303
  Headers:
    Location: http://example.org/final-url

**Comment by benvolioT** [comment]
More comments from @mmorga :  

I know I mentioned the use of Web Link header
over Location header for the status document (per the discussion thread I
linked in another message), but I wouldn¹t have a problem using the
Location header pointing to the status monitor since that is better
understood and more common.

I prefer 303 See Other over 307/308 redirect.

**Comment by mmorga** [comment]
Thanks @benvolioT!

**Comment by benvolioT** [comment]
Eric: I suggest addressing the "link header vs location header" question in an FAQ on the document.

**Comment by benvolioT** [comment]
The prefer header from [RFC 7240](http://tools.ietf.org/html/rfc7240) is interesting. I don't think I've seen that before.

Another approach to consider is that asynchronous requests may be addressed to a different resource.  So for example:

sync: POST http://example.com/ticket
async: POST http://example.com/ticket-job

names from my examples aren't great, but the idea is that the noun we're identifying for the async resource is a different kind of thing.  Just a suggestion.
